src/fact_checker/fact_info.rs

```diff
@@ -0,0 +1,72 @@
+//! Fact info structure and helpers.
+//!
+//! Port of https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/cairo/bootloaders/generate_fact.py
+
+use alloy::primitives::{keccak256, B256};
+use cairo_vm::{
+    program_hash::compute_program_hash_chain,
+    types::{builtin_name::BuiltinName, relocatable::MaybeRelocatable},
+    vm::runners::cairo_pie::CairoPie,
+    Felt252,
+};
+use starknet::core::types::FieldElement;
+
+use super::{
+    error::FactCheckerError,
+    fact_node::generate_merkle_root,
+    fact_topology::{get_fact_topology, FactTopology},
+};
+
+pub const BOOTLOADER_VERSION: usize = 1;
+
+pub struct FactInfo {
+    pub program_output: Vec<Felt252>,
+    pub fact_topology: FactTopology,
+    pub fact: B256,
+}
+
+pub fn get_fact_info(cairo_pie: &CairoPie, program_hash: Option<FieldElement>) -> Result<FactInfo, FactCheckerError> {
+    let program_output = get_program_output(cairo_pie)?;
+    let fact_topology = get_fact_topology(cairo_pie, program_output.len())?;
+    let program_hash = match program_hash {
+        Some(hash) => hash,
+        None => compute_program_hash_chain(&cairo_pie.metadata.program, BOOTLOADER_VERSION)?,
+    };
+    let output_root = generate_merkle_root(&program_output, &fact_topology)?;
+    let fact = keccak256([program_hash.to_bytes_be(), *output_root.node_hash].concat());
+    Ok(FactInfo { program_output, fact_topology, fact })
+}
+
+pub fn get_program_output(cairo_pie: &CairoPie) -> Result<Vec<Felt252>, FactCheckerError> {
```

Comment:
we should ideally have test cases for this, what do you think is the best approach to test this? can we utilise mainnet data somehow? we can do it in a new PR

Reply:
I planned to take/generate test vectors using cairo-lang

Reply:
Awesome, let's do a separate PR. Can you create an issue if possible?